File: 백준/Gold/2636. 치즈/치즈.cc
#include<iostream>
#include<string.h>
#include<utility>
#include<queue>
using namespace std;
typedef pair<int, int>p;
queue<p>q1;
queue<p>q2;
queue<p>rq;
int a[105][105], dx[4] = { 0,0,-1,1 }, dy[4] = { 1,-1,0,0 }, time;
bool v[105][105];
int ans, n, m, cnt, xx;
void del() {
	xx = 0;
	while (!rq.empty()) {
		a[rq.front().first][rq.front().second] = 0;
		ans--;
		rq.pop();
		xx++;
	}
}
void bfs1(int x, int y) {
	q1.push({ x,y });
	v[x][y] = 1;
	while (!q1.empty()) {
		p now = q1.front();
		q1.pop();
		for (int i = 0; i < 4; i++) {
			int nx = now.first + dx[i], ny = now.second + dy[i];
			if (nx < 0 || nx >= n || ny < 0 || ny >= m)continue;
			if (!v[nx][ny] && !a[nx][ny]) {
				q1.push({ nx,ny });
				v[nx][ny] = 1;
			}
			else if (!v[nx][ny] && a[nx][ny]) {
				v[nx][ny] = 1;
				rq.push({ nx,ny });
			}
		}
	}
}
void bfs2(int x, int y) {
	q2.push({ x,y });
	v[x][y] = 1;
	while (!q2.empty()) {
		p now = q2.front();
		q2.pop();
		for (int i = 0; i < 4; i++) {
			int nx = now.first + dx[i], ny = now.second + dy[i];
			if (nx < 0 || nx >= n || ny < 0 || ny >= m)continue;
			if (!v[nx][ny] && a[nx][ny]) {
				q2.push({ nx,ny });
				v[nx][ny] = 1;
			}
		}
	}
}
int main() {
	scanf("%d %d", &n, &m);
	for (int i = 0; i < n; i++)
		for (int j = 0; j < m; j++) {
			scanf("%d", &a[i][j]);
			if (a[i][j])ans++;
		}
	while (1) {
		time++;
		cnt = 0;
		bfs1(0, m / 2);
		del();
		memset(v, 0, sizeof(v));
		for (int i = 0; i < n; i++)
			for (int j = 0; j < m; j++)
				if (!v[i][j] && a[i][j]) {
					bfs2(i, j);
					cnt++;
				}
		if (!cnt) {
			printf("%d\n", time);
			printf("%d", xx);
			return 0;
		}
		memset(v, 0, sizeof(v));
	}
}

File: 백준/Gold/2636. 치즈/치즈.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.*;

public class Main {
    static class Pair {
        int x;
        int y;

        public Pair(int x, int y) {
            this.x = x;
            this.y = y;
        }
    }

    static Queue<Pair> q1 = new LinkedList<>();
    static Queue<Pair> q2 = new LinkedList<>();
    static Queue<Pair> rq = new LinkedList<>();
    static int[][] arr;
    static int[] dx = {0, 0, -1, 1};
    static int[] dy = {1, -1, 0, 0};
    static boolean[][] check;
    static int ans, n, m, cnt, xx, time;

    static void del() {
        xx = 0;
        while (!rq.isEmpty()) {
            arr[rq.peek().x][rq.peek().y] = 0;
            ans--;
            rq.remove();
            xx++;
        }
    }

    static void bfs1(int x, int y) {
        q1.add(new Pair(x, y));
        check[x][y] = true;
        while (!q1.isEmpty()) {
            Pair now = q1.peek();
            q1.remove();
            for (int i = 0; i < 4; i++) {
                int nx = now.x + dx[i], ny = now.y + dy[i];
                if (nx < 0 || nx >= n || ny < 0 || ny >= m) continue;
                if (!check[nx][ny] && arr[nx][ny] == 0) {
                    q1.add(new Pair(nx, ny));
                    check[nx][ny] = true;
                } else if (!check[nx][ny] && arr[nx][ny] == 1) {
                    check[nx][ny] = true;
                    rq.add(new Pair(nx, ny));
                }
            }
        }
    }

    static void bfs2(int x, int y) {
        q2.add(new Pair(x, y));
        check[x][y] = true;
        while (!q2.isEmpty()) {
            Pair now = q2.peek();
            q2.remove();
            for (int i = 0; i < 4; i++) {
                int nx = now.x + dx[i], ny = now.y + dy[i];
                if (nx < 0 || nx >= n || ny < 0 || ny >= m) continue;
                if (!check[nx][ny] && arr[nx][ny] == 1) {
                    q2.add(new Pair(nx, ny));
                    check[nx][ny] = true;
                }
            }
        }
    }

    static void input() throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st;
        st = new StringTokenizer(br.readLine());
        n = Integer.parseInt(st.nextToken());
        m = Integer.parseInt(st.nextToken());
        arr = new int[n][m];
        for (int i = 0; i < n; i++) {
            st = new StringTokenizer(br.readLine());
            for (int j = 0; j < m; j++) {
                arr[i][j] = Integer.parseInt(st.nextToken());
            }
        }
    }

    public static void main(String[] args) throws IOException {
        input();

        while (true) {
            time++;
            cnt = 0;
            check = new boolean[n][m];
            bfs1(0, m / 2);
            del();
            check = new boolean[n][m];
            for (int i = 0; i < n; i++)
                for (int j = 0; j < m; j++)
                    if (!check[i][j] && arr[i][j] == 1) {
                        bfs2(i, j);
                        cnt++;
                    }
            if (cnt == 0) {
                System.out.println(time);
                System.out.println(xx);
                return;
            }
        }
    }
}